pilot/pkg/config/kube/ingress/controller.go

```diff
@@ -170,19 +170,31 @@ func NewController(client kube.Client, meshWatcher mesh.Holder,
 	ingressInformer.AddEventHandler(
 		cache.ResourceEventHandlerFuncs{
 			AddFunc: func(obj interface{}) {
-				q.Push(func() error {
+				key, err := keyFunc(obj)
+				if err != nil {
+					return
+				}
+				q.Push(key, func() error {
```

Comment:
I don't think this is safe. 

There are three queue models:
* Enqueue tasks. This just means we naively process each event, in order. This existing Istio impl
* Enqueue keys. When we process the key, we _reconcile_ the state based on the current state of the world. This is k8s impl
* Enqueue keys and tasks. This is the PR here. This will not work because we are mixing the two states.

For example, say I have:
1. Update ingress update, from class=istio to class=ignored.
2. Delete ingress.

Before, we would notice we stopped watching the class, and trigger an update. Then later the delete comes in, and we see the oldObj is class=ignore, so we do nothing.

With this new change, the delete event will override the update event. This causes us to only process a "delete class=ignored" event, which we do nothing for. This will lead to stale config.

While its valid to enqueue tasks and its valid to enqueue keys, they are not drop in replacements. The actual queue processors need to behave quite differently for the different cases.

I am fine with refactoring our queue usage, but I think it should be incremental and we should not mix keys+tasks.

Reply:
I donot think the example you put up is a problem. For an ingress, we only cares about whether we should process it and what gateway/vs it can be converted.

Changing the class --> delete it, should result in no gateway/vs associated with it. 

I think with/without this pr, the ingress event handling is both not quite right.


> Update ingress update, from class=istio to class=ignored. 

will not trigger vs/gateway delete event, i think we need an separate fix

Reply:
Other than that, i could not find why cannot we mix keys+tasks. The task is just to get the latest object since we are unifying the queue reconcile.

Reply:
> Other than that, i could not find why cannot we mix keys+tasks. The task is just to get the latest object since we are unifying the queue reconcile.

I suppose we can but if the task isn't always the same its probably a bug... so seems like we should associate the task with the queue instance, not the task. This is more like how k8s queue works?